File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.cc
#include<iostream>
#include <vector>
#include <queue>
#include <cstring>
#include <tuple>
#include <algorithm>

using namespace std;
int N,M,K;
int arr[51][51];
vector<tuple<int, int, int>> v;
bool selected[6];
int res = 1e9;
void move(int a,int b,int c) {
    //cout << a << " " << b << " " << c << endl;
    if (c == 0)
        return;
    int tmp = arr[a - c][b - c];
    for (int i = a - c; i < a + c; i++) {
        //cout << arr[i][b - c] << " " << arr[i + 1][b - c] << endl;
        arr[i][b-c] = arr[i + 1][b-c];
    }
    for (int j = b - c; j < b + c; j++) {
        arr[a + c][j] = arr[a + c][j + 1];
    }
    for (int i = a + c; i > a - c; i--) {
        arr[i][b + c] = arr[i - 1][b + c];
    }
    for (int j = b + c; j > b - c; j--) {
        arr[a - c][j] = arr[a - c][j - 1];
    }
    arr[a - c][b - c + 1] = tmp;
    
    move(a,b,c-1);
    
}
void cal() {
    for (int i = 0; i < N; i++) {
            int sum = 0;
            for (int j = 0; j < M; j++) {
                //cout << arr[i][j] << " ";
                sum += arr[i][j];
            }
            //cout << endl;
            res = min(sum, res);
        }
}
void select_order(int n) {
    if (n == K) {
        cal();
        return;
    }
    for (int i = 0; i < K; i++) {
        if (!selected[i]) {
            selected[i] = true;
            int tmp[51][51];
            copy(&arr[0][0], &arr[0][0] + 51*51, &tmp[0][0]);
            move(get<0>(v[i]),get<1>(v[i]),get<2>(v[i]));
            select_order(n + 1);
            copy(&tmp[0][0], &tmp[0][0] + 51*51, &arr[0][0]);
            
            selected[i] = false;
        }

    }
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    cin >> N >> M >> K;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> arr[i][j];
        }
    }
    for (int i = 0; i < K; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        v.push_back(make_tuple(a-1, b-1, c));
    }
    select_order(0);
    
    
    cout << res << endl;
    return 0;
}

File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M, K;
    static int[][] arr = new int[51][51];
    static ArrayList<Tuple> v = new ArrayList<>();
    static boolean[] selected = new boolean[6];
    static int res = 1000000000;
    
    public static void move(int a, int b, int c) {
        if (c == 0) return;
        int tmp = arr[a - c][b - c];
        for (int i = a - c; i < a + c; i++) {
            arr[i][b-c] = arr[i + 1][b-c];
        }
        for (int j = b - c; j < b + c; j++) {
            arr[a + c][j] = arr[a + c][j + 1];
        }
        for (int i = a + c; i > a - c; i--) {
            arr[i][b + c] = arr[i - 1][b + c];
        }
        for (int j = b + c; j > b - c; j--) {
            arr[a - c][j] = arr[a - c][j - 1];
        }
        arr[a - c][b - c + 1] = tmp;

        move(a, b, c - 1);
    }
    
    public static void cal() {
        for (int i = 0; i < N; i++) {
            int sum = 0;
            for (int j = 0; j < M; j++) {
                sum += arr[i][j];
            }
            res = Math.min(sum, res);
        }
    }
    
    public static void select_order(int n) {
        if (n == K) {
            cal();
            return;
        }
        for (int i = 0; i < K; i++) {
            if (!selected[i]) {
                selected[i] = true;
                int[][] tmp = new int[51][51];
                for (int j = 0; j < 51; j++) {
                    tmp[j] = Arrays.copyOf(arr[j], arr[j].length);
                }
                move(v.get(i).a, v.get(i).b, v.get(i).c);
                select_order(n + 1);
                for (int j = 0; j < 51; j++) {
                    arr[j] = Arrays.copyOf(tmp[j], tmp[j].length);
                }
                selected[i] = false;
            }
        }
    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken()) - 1;
            int b = Integer.parseInt(st.nextToken()) - 1;
            int c = Integer.parseInt(st.nextToken());
            v.add(new Tuple(a, b, c));
        }
        select_order(0);
        System.out.println(res);
    }
    
    static class Tuple {
        int a, b, c;
        Tuple(int a,int b,int c){
            this.a=a;
            this.b=b;
            this.c=c;
        }
    }
}